src/Dom/Document.php

```diff
@@ -969,6 +978,28 @@ private function get_mustache_tag_placeholders() {
 		return $placeholders;
 	}
 
+	/**
+	 * Covert the emoji AMP symbol (⚡) into pure text.
+	 *
+	 * The emoji symbol gets stripped by DOMDocument::loadHTML().
+	 *
+	 * @param string $source Source HTML string to convert the emoji AMP symbol in.
+	 * @return string Adapted source HTML string.
+	 */
+	private function convert_amp_emoji_attribute( $source ) {
+		return preg_replace( '/(<html [^>]*?)⚡([^\s^>]*)/i', '\1' . self::EMOJI_AMP_ATTRIBUTE . '="\2"', $source, 1 );
+	}
+
+	/**
+	 * Restore the emoji AMP symbol (⚡) from its pure text placeholder.
+	 *
+	 * @param string $html HTML string to restore the AMP emoji symbol in.
+	 * @return string Adapted HTML string.
+	 */
+	private function restore_amp_emoji_attribute( $html ) {
+		return preg_replace( '/(<html [^>]*?)' . preg_quote( self::EMOJI_AMP_ATTRIBUTE, '/' ) . '="([^"]*)"/i', '\1⚡\2', $html, 1 );
+	}
```

Comment:
I'm trying to build this so that we can use the language-independent testsuite provided at https://github.com/ampproject/amp-toolbox/tree/master/packages/optimizer/spec/transformers/valid

We can of course either ignore these tests, or modify them on-the-fly to adapt them to our specific requirements. But I thought it might be worthwhile to stick as closely to the testable spec as possible within the PHP libraries...?